### PR TITLE
REGR: setitem writing into RangeIndex instead of creating a copy

### LINKED ISSUE #47128: BUG: RangeIndex problem
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
A = pd.DataFrame({'a':[False,True]}) # index=[0,1] vs index=range(2)
A['i'] = A.index
A.loc[A['a'],'i'] = -2
print(A.loc[A['a'],:])
```

### Issue Description

when assign index(RangeIndex) to new column then change this column by loc, index will change unexpectedly. this issue exists in 1.4.x but not in 1.1.1

<output>
<html>
<body>

  | a | i
-- | -- | --
-2|True | -2

</body>
</html>
</output>

### Expected Behavior

<html>
<body>

  | a | i
-- | -- | --
1|True | -2

</body>
</html>

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : bb1f651536508cdfef8550f93ace7849b00046ee
python           : 3.9.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-110-generic
Version          : #124-Ubuntu SMP Thu Apr 14 19:46:19 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.0
numpy            : 1.22.2
pytz             : 2022.1
dateutil         : 2.8.2
pip              : 21.2.3
setuptools       : 49.2.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.8.0
html5lib         : None
pymysql          : None
psycopg2         : 2.9.1
jinja2           : 3.1.2
IPython          : 7.33.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.4.3
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.3
sqlalchemy       : 1.4.23
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #47128 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/construction.py [hunk 1]
```diff
@@ -531,7 +531,7 @@ def sanitize_array(
         dtype = dtype.numpy_dtype
 
     # extract ndarray or ExtensionArray, ensure we have no PandasArray
-    data = extract_array(data, extract_numpy=True)
+    data = extract_array(data, extract_numpy=True, extract_range=True)
```
Comment (phofl) [inline pandas/core/construction.py:534]: I think this is the correct place to handle RangeIndexes, but this leaves us in the same position with other array types, like dask arrays, hence the copy below

File: pandas/core/construction.py [hunk 2]
```diff
@@ -611,6 +611,8 @@ def sanitize_array(
         if hasattr(data, "__array__"):
             # e.g. dask array GH#38645
             data = np.asarray(data)
+            if copy:
```
Comment (phofl) [inline pandas/core/construction.py:614]: Is there an efficient way to check if asarray already made a copy?

File: doc/source/whatsnew/v1.4.3.rst [hunk 3]
```diff
@@ -15,6 +15,7 @@ including other versions of pandas.
 Fixed regressions
 ~~~~~~~~~~~~~~~~~
 - Fixed regression in :meth:`DataFrame.replace` when the replacement value was explicitly ``None`` when passed in a dictionary to ``to_replace`` also casting other columns to object dtype even when there were no values to replace (:issue:`46634`)
+- Fixed regression in :meth:`DataFrame.loc.__setitem__` updating :class:`RangeIndex` when index was set as new column and column was updated afterwards (:issue:`47128`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.4.3.rst:18]: ```:meth:`DataFrame.loc.__setitem__` ``` doesn't render a link.

```suggestion
- Fixed regression when setting values with  :meth:`DataFrame.loc` updating :class:`RangeIndex` when index was set as new column and column was updated afterwards (:issue:`47128`)
```

Comment (phofl) [reply doc/source/whatsnew/v1.4.3.rst:18]: Ah thx.

File: pandas/core/construction.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/construction.py:614]: Maybe just change the call above to `data = np.array(data, copy=copy)`?

Comment (phofl) [reply pandas/core/construction.py]: Thx, good point

**Review by mroeschke — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl very nice!

**Comment by jreback** [comment]
@meeseeksdev backport 1.4.x

**Pull request merged**